### DISCUSSION TITLE
How to trim (remove) white border around reactions in picture

### DISCUSSION THREAD

**Discussion by ddsneo4j** [rdkit/rdkit#5966] [Development]
Hi,

 I know there is a similar thread (#5879) around removing white border around **structure** pictures, but unfortunately that does not work for allmost all of my **reaction** pictures.

I am using the following code:

        rxn = AllChem.ReactionFromRxnBlock(r)
    
        d2d = Draw.MolDraw2DCairo(-1,-1)
        d2d.drawOptions().useBWAtomPalette()
        # use highlighting to avoid images with atom ids, highlighting set to white to be invisible
        d2d.DrawReaction(rxn,highlightByReactant=True,highlightColorsReactants=[(1,1,1),] )
        
        d2d.FinishDrawing()

        png = d2d.GetDrawingText()
        
        name = k[len('Short-RInChIKey='):]
        filename_trim = 'C:\\images_trim\\' + name + '.png'
        with open(filename_trim, 'wb') as f:
            f.write(png)

The generated reaction picture looks like this, i.e. the white canvas around it was not removed:

![image](https://user-images.githubusercontent.com/42145906/212115592-da8717e5-47be-4a63-b5dc-36fdc8c27b23.png)

Please find a screenshot and the used rxn attached.

Any help is highly appreciated.

Thanks,
Dan

[r1.rxn.txt](https://github.com/rdkit/rdkit/files/10403355/r1.rxn.txt)

**Comment by DavidACosgrove** [comment]
Firstly, the reaction you have provided is not the one drawn.  That doesn't really matter in this instance, but it wasn't overly helpful, neither was the fact that your code snippet didn't run without further editing.  I have amended it to:

```
run = AllChem.ReactionFromRxnBlock(r)

d2d = rdMolDraw2D.MolDraw2DCairo(-1,-1)

rdMolDraw2D.SetDarkMode(d2d)
d2d.drawOptions().padding = 0.0
d2d.DrawReaction(rxn,highlightByReactant=True,highlightColorsReactants=[(1,0,0),] )
    
d2d.FinishDrawing()

png = d2d.GetDrawingText()
    
with open('test.png', 'wb') as f:
    f.write(png)
```

With 2022.09.3, I got this picture:
![image](https://user-images.githubusercontent.com/9198870/212483619-d570c128-24ca-4bc0-8e8c-6c7382f5e34c.png)
I have used dark mode and red highlights so you can see that the picture is going to the edge of the image/canvas. in your image, the white highlights on a white background are bound to produce the appearance of a border, because you can't see the highlight circles. I don't understand the comment in your code about atom ids in the image since these are not produced by default.  Setting the draw options padding to 0.0 is important for not having any border.

The highlight circles are clipped at the edge of the above image in places and a bit distant from it in others due to a bug in the way the extremes of the circles were calculated, which is fixed and will be in the upcoming patch release.  With the current master branch, the image I get is:

![test](https://user-images.githubusercontent.com/9198870/212483798-c6ffe37b-5785-4631-b536-332c270ca296.png)
Which if anything is still cropping the products a bit but certainly doesn't have a border.

**Reply by ddsneo4j** [reply]
Thanks @DavidACosgrove for your effort and sorry for the inconsistency between depicted image and rxn. However, in my version 2022.09.1 the atom ids are actually produced by default. That is why highlighting was used as a workaround. I will try the version you used and see if that issue disappears and thus helpf cropping. Even with padding=0.0 there is a tendency that the bottom and right hand side of the reaction picture are less cropped than the other sides. Ideally, a border of 5 pixels would be fine, but playing with the padding parameter works great for some reactions but not at all for others. I am afraid I can't give you our real data, so I have to fake it for the public as you may have already guessed. As I see this procedure of cropping does not work overly consistent applied to our dataset, which makes it really problematic. I will wait for the update you mentioned and try again with that.

**Reply by ddsneo4j** [reply]
@DavidACosgrove I updated rdkit to 2022.09.3, but without the highlighting workaround the atom ids are still produced by default. Is this really not the case for you? What's different then?

**Comment by DavidACosgrove** [comment]
No, I have no atom indices. Can you post the code you use and the image without white highlights so I can see exactly what you mean. Especially it would be useful to see what drawing options you are using as well as setting padding to 0.

**Reply by ddsneo4j** [reply]
@DavidACosgrove I am using the following code:

        rxn = AllChem.ReactionFromRxnBlock(r)
        d2d = Draw.MolDraw2DCairo(-1,-1)
        d2d.drawOptions().useBWAtomPalette()
        d2d.drawOptions().padding = 0.0
        d2d.DrawReaction(rxn)   
        d2d.FinishDrawing()
        png = d2d.GetDrawingText()
        
        filename_trim = 'C:\\images_trim\\test.png'
        with open(filename_trim, 'wb') as f:
            f.write(png)

And the png looks like this:

![test](https://user-images.githubusercontent.com/42145906/213715673-db10a40c-c664-415a-9f33-e4de6b9c070e.png)
[r1.rxn.txt](https://github.com/rdkit/rdkit/files/10466917/r1.rxn.txt)

**Reply by DavidACosgrove** [reply]
Thanks, @ddsneo4j, I now understand the issue.  Those are the atom map numbers, not the atom ids, though that is a distinction that perhaps isn't clearly made in the documentation, such as it is.  The atom map numbers show where each atom in the reactants end up in the products.  The atom highlights aren't hiding the map numbers as such - if you use the highlightByReactant option, the atom maps aren't drawn at all.  To not have them drawn generally, you can set them to 0 before drawing:
```
rxn = AllChem.ReactionFromRxnBlock(r)
for prod in rxn.GetProducts():
    for atom in prod.GetAtoms():
        atom.SetAtomMapNum(0)
for reagent in rxn.GetReactants():
    for atom in reagent.GetAtoms():
        atom.SetAtomMapNum(0)
    
d2d = rdMolDraw2D.MolDraw2DCairo(-1,-1)

d2d.drawOptions().useBWAtomPalette()
d2d.drawOptions().padding = 0.0
d2d.DrawReaction(rxn)    
d2d.FinishDrawing()
png = d2d.GetDrawingText()
    
with open('test.png', 'wb') as f:
    f.write(png)
```
Which produces:
![test](https://user-images.githubusercontent.com/9198870/213858427-57d9500e-bd70-4e33-bad7-031e44a8206c.png)
with 2022.09.4.
@greglandrum do you think it would be worth adding a drawOption to not draw the atom maps?

**Reply by ddsneo4j** [reply]
Thank you so much, @DavidACosgrove, it works much better now! 
As I am processing millions of reactions, it would be great to have the drawOption that you suggested to not draw the atom maps because iterating over the atom map numbers slows performance down.

In addition, would it be possible to add a n-pixel white border? Changing padding from 0.0 to anything else created a much bigger border for me.

Do you know why in your picture the bonds get that bold? I see this issue as well on a lot of reaction PNGs that makes them impossible to read.

Thanks.

**Reply by ddsneo4j** [reply]
@greglandrum Any chance to get that drawOption mentioned by @DavidACosgrove implemented soon?
